### PR TITLE
DDP-8465 | Onc History Detail

### PR DESCRIPTION
**Pull request by ScalaChamp**
## Context

Ticket: https://broadinstitute.atlassian.net/browse/DDP-8465

Problem: In ElasticSearch `dsm.participant` holds data for `ddp_participant` and `ddp_participant_record` tables. Since we are using `@TableName` annotations which add the metadata to the DTO classes we end up with computing the incorrect `primary key` and DB operations fail.

Steps to reproduce:
   - pick any study
   - find `enrolled` participant with no records for `Onc History`
   - navigate to participant page
   - navigate to `Onc History`
   - input a value in some field so that the client sends patch request
   - backend throws a RuntimeException with the message: `Id and parentId was null`.

Solution: In case of `r` (ddp_participant_record) and `p` (ddp_participant) the primary key should be `participant_id` and not `participant_record_id` in order for DB operations to run correctly.

Details:
  -  add `DBElementBuilder.java` interface for exposing the API for building `DBElement` instances from `NameValue#name`
  -  create two concrete implementations for `DBElementBuilder.java`:
      -  `DefaultDBElementBuilder.java` that simply delegates to `PatchUtil`
      -  `ParticipantRecordDBElementBuilder.java` which calls the super and then changes the primary key to `participant_id`
  -  establish a new factory class called `ExistingRecordPatchFactory.java` which can produce either of two patch instances:
      -  `ExistingParticipantRecordPatch.java` for updating `participant_record`
      -  `ExistingRecordPatch.java` for updating any other record such as `medical_record`, `onc_history_detail` etc..
  -  `ExistingParticipantRecordPatch.java` will use `ParticipantRecordDBElementBuilder.java` as a dependency for building `DBElement`
  -  `ExistingRecordPatch.java` will use the `DefaultDBElementBuilder.java` for building `DBElement`
  -    cover the majority of cases with unit tests.
      
_Say a few words about the bigger picture beyond this PR. How do your changes improve life for our users? Link directly to the jira ticket (just type "DDP-..." and it'll automatically link) to help your teammates quickly get up to speed on the motivation for these changes._

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/DefaultDBElementBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import org.broadinstitute.dsm.db.structure.DBElement;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class DefaultDBElementBuilderTest {
+
+    @BeforeClass
+    public static void initializeDB() {
+        TestHelper.setupDB();
```
Comment (gmakharat) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/DefaultDBElementBuilderTest.java]: is it really necessary? I don't see any test which requires any DB connection

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/ParticipantRecordDBElementBuilderTest.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import org.broadinstitute.dsm.db.structure.DBElement;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class ParticipantRecordDBElementBuilderTest {
+
+    @BeforeClass
+    public static void initializeDB() {
+        TestHelper.setupDB();
```
Comment (gmakharat) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/ParticipantRecordDBElementBuilderTest.java]: the same

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/DefaultDBElementBuilderTest.java (hunk 1 above)
Comment (ScalaChamp) [reply pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/DefaultDBElementBuilderTest.java]: @gmakharat Thank you for noticing it, it's redundant, just removing it.

**Review by SimoneMaiwald — DISMISSED** [review]

File: pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/DefaultDBElementBuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    @Test
+    public void fromNonExistingName() {
+        var name = "We love DSM";
+        var builder = new DefaultDBElementBuilder();
+
+        Assert.assertThrows(RuntimeException.class, () -> builder.fromName(name));
+    }
+
+}
```
Comment (SimoneMaiwald) [inline pepper-apis/dsm-core/src/test/java/org/broadinstitute/dsm/model/patch/DefaultDBElementBuilderTest.java]: checkstyle!!!

**Comment by SimoneMaiwald** [comment]
before you merge it, please check your code for checkstyle again

**Review by gmakharat — APPROVED** [review]

**Pull request merged**